File: chaizup_toc/chaizup_toc/report/production_priority_board/production_priority_board.js
// Production Priority Board — Client-side filters + action buttons

frappe.query_reports["Production Priority Board"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
        },
        {
            fieldname: "buffer_type",
            label: __("Buffer Type"),
            fieldtype: "Select",
            options: "\nFG\nSFG\nRM\nPM",
            description: "FG=Finished | SFG=Semi-Finished | RM=Raw Material | PM=Packaging",
        },
        {
            fieldname: "warehouse",
            label: __("Warehouse"),
            fieldtype: "Link",
            options: "Warehouse",
        },
        {
            fieldname: "zone",
            label: __("Zone Filter"),
            fieldtype: "Select",
            options: "\nGreen\nYellow\nRed\nBlack",
            description: "F3: Filter by buffer zone. Leave blank for all.",
        },
        {
            fieldname: "item_code",
            label: __("Item"),
            fieldtype: "Link",
            options: "Item",
            get_query: function() {
                return { filters: { custom_toc_enabled: 1 } };
            },
        },
    ],

    onload(report) {
        // ── ACTION BUTTON: Generate Material Requests NOW ──
        report.page.add_inner_button(__("🔴 Generate MRs Now"), function() {
            frappe.confirm(
                "<b>Run TOC Buffer Analysis Now?</b><br><br>" +
                "This will calculate all buffers and create Material Requests " +
                "for Red/Yellow zone items immediately.<br><br>" +
                "<small>Formula: F4 Order Qty = Target Buffer − Inventory Position</small>",
                function() {
                    frappe.call({
                        method: "chaizup_toc.api.toc_api.trigger_manual_run",
                        freeze: true,
                        freeze_message: __("Running TOC Buffer Analysis..."),
                        callback(r) {
                            if (r.message) {
                                let msg = `<b>${r.message.created}</b> Material Requests created.`;
                                if (r.message.material_requests && r.message.material_requests.length) {
                                    msg += "<br><br>MRs: " + r.message.material_requests.map(
                                        mr => `<a href="/app/material-request/${mr}">${mr}</a>`
                                    ).join(", ");
                                }
                                frappe.msgprint({title: "TOC Run Complete", message: msg, indicator: "green"});
                                report.refresh();
                            }
                        }
                    });
                }
            );
        }, __("Actions"));

        // ── ACTION BUTTON: Apply DAF (Seasonal Adjustment) ──
        report.page.add_inner_button(__("📅 Apply DAF"), function() {
            let d = new frappe.ui.Dialog({
                title: "Apply Demand Adjustment Factor (F6)",
                fields: [
                    {fieldtype: "Float", fieldname: "daf", label: "DAF Multiplier",
                     default: 1.0, reqd: 1,
                     description: "F6: Adjusted Buffer = Target × DAF. Diwali=1.6, Summer=0.7, Normal=1.0"},
                    {fieldtype: "Select", fieldname: "preset", label: "Or Select Preset",
                     options: "\nDiwali (1.6x)\nTrade Promotion (1.8x)\nYear-End (1.3x)\nMonsoon (0.85x)\nSummer (0.7x)\nNormal (1.0x)",
                     change: function() {
                         let map = {"Diwali (1.6x)":1.6, "Trade Promotion (1.8x)":1.8,
                                    "Year-End (1.3x)":1.3, "Monsoon (0.85x)":0.85,
                                    "Summer (0.7x)":0.7, "Normal (1.0x)":1.0};
                         let v = map[d.get_value("preset")];
                         if (v) d.set_value("daf", v);
                     }},
                    {fieldtype: "Data", fieldname: "event_name", label: "Event Name",
                     description: "Descriptive name (e.g. 'Diwali 2026')"},
                ],
                primary_action_label: "Apply to All Buffers",
                primary_action(values) {
                    frappe.call({
                        method: "chaizup_toc.api.toc_api.apply_global_daf",
                        args: {daf_value: values.daf, event_name: values.event_name},
                        freeze: true,
                        callback(r) {
                            if (r.message) {
                                frappe.msgprint(`DAF ${values.daf}x applied to ${r.message.updated_rules} buffer rules.`);
                                report.refresh();
                            }
                        }
                    });
                    d.hide();
                }
            });
            d.show();
        }, __("Actions"));

        // ── ACTION BUTTON: Reset DAF ──
        report.page.add_inner_button(__("↩️ Reset DAF"), function() {
            frappe.confirm("Reset DAF to 1.0x (Normal Operations) for all buffers?", function() {
                frappe.call({
                    method: "chaizup_toc.api.toc_api.reset_global_daf",
                    freeze: true,
                    callback(r) {
                        frappe.msgprint("DAF reset to 1.0x (Normal Operations).");
                        report.refresh();
                    }
                });
            });
        }, __("Actions"));

        // ── ACTION BUTTON: Open TOC Guide ──
        report.page.add_inner_button(__("📖 TOC Guide"), function() {
            window.open("/toc-guide", "_blank");
        });
    },

    // Row formatting — color by zone
    formatter(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (data && column.fieldname === "zone") {
            let colors = {Green:"#27AE60", Yellow:"#F39C12", Red:"#E74C3C", Black:"#2C3E50"};
            let bgs = {Green:"#D5F5E3", Yellow:"#FEF9E7", Red:"#FADBD8", Black:"#D5D8DC"};
            let c = colors[data.zone] || "#7F8C8D";
            let bg = bgs[data.zone] || "#FFF";
            value = `<span style="background:${bg};color:${c};padding:3px 10px;border-radius:12px;font-weight:bold;font-size:11px">${data.zone}</span>`;
        }

        if (data && column.fieldname === "bp_pct" && data.bp_pct >= 67) {
            value = `<span style="color:#E74C3C;font-weight:bold">${data.bp_pct}%</span>`;
        } else if (data && column.fieldname === "bp_pct" && data.bp_pct >= 33) {
            value = `<span style="color:#F39C12;font-weight:bold">${data.bp_pct}%</span>`;
        }

        if (data && column.fieldname === "order_qty" && data.order_qty > 0) {
            value = `<span style="color:#E67E22;font-weight:bold">${frappe.format(data.order_qty, {fieldtype:'Float'})}</span>`;
        }

        if (data && column.fieldname === "zone_action") {
            let c = {Green:"#27AE60", Yellow:"#F39C12", Red:"#E74C3C", Black:"#E74C3C"}[data.zone] || "#7F8C8D";
            value = `<span style="color:${c};font-weight:bold">${data.zone_action}</span>`;
        }

        return value;
    },
};

File: chaizup_toc/chaizup_toc/report/production_priority_board/production_priority_board.py
"""
Production Priority Board — Script Report
===========================================
THE daily report. Every morning at 7:00 AM the supervisor opens this.
Shows ALL TOC-enabled items sorted by Buffer Penetration % — highest urgency first.
No item-type filter — every item is shown regardless of category.

Formulas shown in every row:
  F2: IP = On-Hand + WIP − Backorders
  F3: BP% = (Target − IP) ÷ Target × 100
  F4: Order Qty = Target − IP
  F5: T/CU for tie-breaking (manufactured items)
"""

import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart(data)
    summary = get_summary(data)

    return columns, data, None, chart, summary


def get_columns():
    return [
        {"fieldname": "rank", "label": _("Rank"), "fieldtype": "Int", "width": 60},
        {"fieldname": "item_code", "label": _("Item Code"), "fieldtype": "Link",
         "options": "Item", "width": 140},
        {"fieldname": "item_name", "label": _("Item Name"), "fieldtype": "Data", "width": 180},
        {"fieldname": "stock_uom", "label": _("UOM"), "fieldtype": "Link",
         "options": "UOM", "width": 60},
        {"fieldname": "buffer_type", "label": _("Mode"), "fieldtype": "Data", "width": 90},
        {"fieldname": "warehouse", "label": _("Warehouse"), "fieldtype": "Link",
         "options": "Warehouse", "width": 140},
        {"fieldname": "target_buffer", "label": _("Target Buffer<br><small>F1: ADU×RLT×VF</small>"),
         "fieldtype": "Float", "width": 120},
        {"fieldname": "on_hand", "label": _("On-Hand"), "fieldtype": "Float", "width": 90},
        {"fieldname": "wip_or_on_order", "label": _("WIP/On-Order"), "fieldtype": "Float", "width": 100},
        {"fieldname": "backorders_or_committed", "label": _("Backorders/<br>Committed"),
         "fieldtype": "Float", "width": 100},
        {"fieldname": "inventory_position", "label": _("IP<br><small>F2: OH+WIP−BO</small>"),
         "fieldtype": "Float", "width": 100},
        {"fieldname": "bp_pct", "label": _("BP%<br><small>F3: (T−IP)÷T</small>"),
         "fieldtype": "Percent", "width": 100},
        {"fieldname": "sr_pct", "label": _("SR%<br><small>IP÷Target</small>"),
         "fieldtype": "Percent", "width": 80},
        {"fieldname": "zone", "label": _("Zone"), "fieldtype": "Data", "width": 80},
        {"fieldname": "order_qty", "label": _("Order Qty<br><small>F4: Target−IP</small>"),
         "fieldtype": "Float", "width": 110},
        {"fieldname": "tcu", "label": _("T/CU<br><small>F5: ₹/min</small>"),
         "fieldtype": "Currency", "width": 100},
        {"fieldname": "zone_action", "label": _("Action"), "fieldtype": "Data", "width": 150},
        {"fieldname": "sfg_message", "label": _("SFG Status"), "fieldtype": "Data", "width": 180},
    ]


def get_data(filters):
    from chaizup_toc.toc_engine.buffer_calculator import calculate_all_buffers

    kwargs = {}
    if filters:
        for k in ("company", "warehouse", "item_code"):
            if filters.get(k):
                kwargs[k] = filters[k]

    buffers = calculate_all_buffers(**kwargs)

    if filters and filters.get("zone"):
        buffers = [b for b in buffers if b["zone"] == filters["zone"]]

    if filters and filters.get("buffer_type"):
        buffers = [b for b in buffers if b.get("buffer_type") == filters["buffer_type"]]

    data = []
    for i, b in enumerate(buffers):
        data.append({
            "rank": i + 1,
            "item_code": b["item_code"],
            "item_name": b["item_name"],
            "stock_uom": b.get("stock_uom", ""),
            "buffer_type": b["buffer_type"],
            "warehouse": b["warehouse"],
            "target_buffer": b["target_buffer"],
            "on_hand": b["on_hand"],
            "wip_or_on_order": b["wip_or_on_order"],
            "backorders_or_committed": b["backorders_or_committed"],
            "inventory_position": b["inventory_position"],
            "bp_pct": b["bp_pct"],
            "sr_pct": b["sr_pct"],
            "zone": b["zone"],
            "order_qty": b["order_qty"],
            "tcu": b["tcu"],
            "zone_action": b["zone_action"],
            "sfg_message": b.get("sfg_status", {}).get("message", "") if b.get("sfg_status") else "",
        })

    return data


def get_chart(data):
    """Zone distribution pie chart."""
    if not data:
        return None

    zone_counts = {"Green": 0, "Yellow": 0, "Red": 0, "Black": 0}
    for d in data:
        zone_counts[d["zone"]] = zone_counts.get(d["zone"], 0) + 1

    return {
        "data": {
            "labels": list(zone_counts.keys()),
            "datasets": [{"name": "Items", "values": list(zone_counts.values())}]
        },
        "type": "pie",
        "colors": ["#27AE60", "#F39C12", "#E74C3C", "#2C3E50"],
        "height": 280,
    }


def get_summary(data):
    """Summary cards at the top of the report."""
    if not data:
        return []

    total = len(data)
    red = len([d for d in data if d["zone"] in ("Red", "Black")])
    yellow = len([d for d in data if d["zone"] == "Yellow"])
    green = len([d for d in data if d["zone"] == "Green"])
    avg_bp = round(sum(d["bp_pct"] for d in data) / total, 1) if total else 0

    return [
        {"value": total, "label": _("Total Buffers"), "datatype": "Int", "indicator": "blue"},
        {"value": red, "label": _("Red/Black (Urgent)"), "datatype": "Int", "indicator": "red"},
        {"value": yellow, "label": _("Yellow (Plan)"), "datatype": "Int", "indicator": "orange"},
        {"value": green, "label": _("Green (OK)"), "datatype": "Int", "indicator": "green"},
        {"value": avg_bp, "label": _("Avg BP%"), "datatype": "Percent", "indicator": "blue"},
    ]
